streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamReduce.java

```diff
@@ -76,8 +76,9 @@ public void process(K key, V value) {
                 }
             }
 
-            // update the store with the new value
-            store.put(key, newAgg);
+            // update the store with the new value if the value is not null
+            if (newAgg != null)
```

Comment:
If the user defined `reduce` or `aggregate` call returns `null`, it means deleting the value for this aggregated key. So we should still put `(key, null)` into the store, which will be handled as a deletion operation.